### PR TITLE
Make postgres index reflection more robust

### PR DESCRIPTION
**Pull request by markstory**
Use postgres' catalog schema more to allow complex index definitions like:

```
CREATE INDEX flags_unaccent_title_idx ON flags USING gin (f_unaccent(title::text) gin_trgm_ops)
```

to be reflected properly.

Refs #7871 
cc @LACabeza

### CODE REVIEW

File: src/Database/Schema/PostgresSchema.php [hunk 1]
```diff
@@ -229,46 +233,54 @@ public function convertIndexDescription(Table $table, $row)
         if ($row['indisunique'] && $type === Table::INDEX_INDEX) {
             $type = Table::CONSTRAINT_UNIQUE;
         }
-        preg_match('/\(([^\)]+?)\s*(?:ASC|DESC)?(?:NULLS FIRST|LAST)?\)/', $row['statement'], $matches);
```
Comment (lorenzo) [inline src/Database/Schema/PostgresSchema.php]: I'm glad this is gone again :D

**Comment by lorenzo** [comment]
Postgres tests failing

**Comment by LACabeza** [comment]
With this sql, when the index is composed by two or more fields, it'll be fetched a row for each of the fields, like.

relname | attname | statement
-------  | ------- | ---------
facts_pkey | source | CREATE UNIQUE INDEX facts_pkey ON facts USING btree (source, code)
facts_pkey | code | CREATE UNIQUE INDEX facts_pkey ON facts USING btree (source, code)

Won't this be a problem? I saw on tests, this may be the problem for testSaveNewErrorCompositeKeyNoIncrement

**Comment by markstory** [comment]
@LACabeza The subselect changes the order of the columns in the index which is also a problem as. That's why I went with the multiple row approach. There is clearly an issue to iron out around composite primary keys though.

**Comment by markstory** [comment]
I think I got this fixed now. I was able to purge out more hacky code too which is nice.

File: tests/TestCase/Database/Schema/PostgresSchemaTest.php [hunk 2]
```diff
@@ -288,7 +288,7 @@ public function testDescribeTable()
                 'precision' => null,
                 'unsigned' => null,
                 'comment' => null,
-                'autoIncrement' => true,
+                'autoIncrement' => false,
```
Comment (markstory) [inline tests/TestCase/Database/Schema/PostgresSchemaTest.php]: The schema definition for this table does not define this column as a serial, and instead only has `BIGINT PRIMARY KEY` which is not a serial type.

Comment (josegonzalez) [inline tests/TestCase/Database/Schema/PostgresSchemaTest.php]: Does that affect how CakePHP will treat the column internally?

Comment (markstory) [inline tests/TestCase/Database/Schema/PostgresSchemaTest.php]: Yes. But the current code results in a more accurate reflection of the schema.

**Pull request merged**